src/mon/MonmapMonitor.h

```diff
@@ -29,16 +30,74 @@
 #include "MonMap.h"
 #include "MonitorDBStore.h"
 
+#include "command_multiplexer.h"
+
+class MMonGetMap;
 class MMonMap;
 class MMonCommand;
 class MMonJoin;
 
+namespace ceph::mon_cmds {
+
+class failure final 
+{
+    const int error_code_;
+    char what_[128] = {};
+
+    public:
+    failure(const int error_code, const std::string msg)
+     : error_code_(error_code)
+    {
+        std::snprintf(what_, sizeof(what_), "error %d: %s", error_code, msg.c_str());
+    };
+
+    virtual ~failure() {}
+
+    public:
+    const int error_code() const noexcept   { return error_code_; }
+    const char *what() const noexcept       { return what_; }
+};
+
+struct mon_cmd_ctx
+{
+ MonOpRequestRef                op;
+ MonmapMonitor&                 monmap_monitor;
+ Monitor&                       mon;
+
+ MMonCommand&                   mmon_command;   // aka "m"
+
+ cmdmap_t                       cmdmap;
+ std::string                    prefix;            // "the command" itself, e.g. "osd set"
+
+ // many functions expect this to be available:
+ std::string                    format;
+ std::unique_ptr<Formatter>     f;
+
+ // output by side-effect:
+ mutable int                    r = 0;      // a result code
+ mutable std::stringstream      ss;         // a string message
+ mutable ceph::bufferlist       rdata;      // a result buffer
+
```

Comment:
I'm not sure how I feel about this.

I understand the reasoning behind it.

It does, as you say, have the advantage that output (or in/out) parameters are very distinctly and definitely marked.

I don't actively DIS-like it.

Reply:
I'm about on the same page: I think it's kind of /ugly/, but also is about the only way to really see and get the compiler to help us understand what things are and are not modified by commands as side effects. It's tough to tell right now. One of my hopes is that after we actually have this command multiplexer (suggestions for other names are welcome, but "dispatcher" I guess is already taken-- "demuxer" is arguably more accurate, but hardly more pleasing) we will have a simpler view of things that lets us see MORE of the underlying patterns.

The goal would be ideally to have a single underlying framework for all of these multiplexing-like tasks. And yes it would also be great if we further subsumed the stuff accomplished with command macros now. I'm actually very keen on doing just that, but also have a little fear that this will start to get large and never actually "in"... :-)